pkg/nsx/services/subnetport/subnetport.go

```diff
@@ -0,0 +1,188 @@
+package subnetport
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"sync"
+
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt/model"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/logger"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+)
+
+var (
+	log                    = logger.Log
+	ResourceTypeSubnetPort = common.ResourceTypeSubnetPort
+	MarkedForDelete        = true
+)
+
+type SubnetPortService struct {
+	common.Service
+	SubnetPortStore *SubnetPortStore
+}
+
+// InitializeSubnetPort sync NSX resources
+func InitializeSubnetPort(service common.Service) (*SubnetPortService, error) {
+	wg := sync.WaitGroup{}
+	wgDone := make(chan bool)
+	fatalErrors := make(chan error)
+
+	wg.Add(1)
+
+	subnetPortService := &SubnetPortService{Service: service}
+
+	subnetPortService.SubnetPortStore = &SubnetPortStore{ResourceStore: common.ResourceStore{
+		Indexer:     cache.NewIndexer(keyFunc, cache.Indexers{common.TagScopeSubnetPortCRUID: indexFunc}),
+		BindingType: model.SegmentPortBindingType(),
+	}}
+
+	go subnetPortService.InitializeResourceStore(&wg, fatalErrors, ResourceTypeSubnetPort, subnetPortService.SubnetPortStore)
+
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		close(fatalErrors)
+		return subnetPortService, err
+	}
+
+	return subnetPortService, nil
+}
+
+func (service *SubnetPortService) CreateOrUpdateSubnetPort(obj *v1alpha1.SubnetPort) error {
+
+	log.Info("Create or update subnetport", "subnetport", obj)
+
+	// TODO: For now, only the parent subnet is supported.
+	specSubnet := obj.Spec.Subnet
+
+	nsxSubnetPath, err := service.GetPathFromSubnet(specSubnet, obj.Namespace, service.Client)
+	if err != nil {
+		log.Error(err, "failed to get NSX resource path from subnet", "subnet", specSubnet, "namespace", obj.Namespace)
+		return err
+	}
+	obj.Status.NSXResourcePath = fmt.Sprintf("%s/ports/%s", nsxSubnetPath, obj.UID)
+
+	nsxSubnetPort, err := service.buildSubnetPort(obj)
+	if err != nil {
+		log.Error(err, "failed to build NSX subnet port")
+		return err
+	}
+
+	paras := strings.Split(nsxSubnetPath, "/")
+	nsxOrgID := paras[2]
+	nsxProjectID := paras[4]
+	nsxVPCID := paras[6]
+	nsxSubnetID := paras[8]
+
+	existingSubnetPort := service.SubnetPortStore.GetByKey(*nsxSubnetPort.Id)
+	isChanged := common.CompareResource(SubnetPortToComparable(existingSubnetPort), SubnetPortToComparable(nsxSubnetPort))
+	if !isChanged {
+		log.Info("NSX subnet port not changed, skipping the update", "nsxSubnetPort.Id", nsxSubnetPort.Id)
+		return nil
+	}
+	err = service.NSXClient.PortClient.Patch(nsxOrgID, nsxProjectID, nsxVPCID, nsxSubnetID, *nsxSubnetPort.Id, *nsxSubnetPort)
```

Comment:
Do we need to check realization status?

Reply:
Yes I think we should, investigating the API...

Reply:
done